test/e2e/external_oidc_test.go

```diff
@@ -187,263 +290,460 @@ func newTestClient(t *testing.T) (*testClient, error) {
+		} else if requireMissing != errors.IsNotFound(err) {
+			return false, fmt.Errorf("apiservice %s wanted missing: %v; got: %v", name, requireMissing, !errors.IsNotFound((err)))
 		}
-	})
-	if err != nil {
-		return
 	}
 
-	return
+	return true, nil
 }
 
-func (tc *testClient) updateAuthForOIDC(ctx context.Context, idpURL, idpName string) (auth *configv1.Authentication, cleanups []func(), err error) {
-	tc.t.Log("will update auth CR for OIDC")
+func (tc *testClient) testOIDCAuthentication(t *testing.T, ctx context.Context, kcClient *test.KeycloakClient, usernameClaim, usernamePrefix string) {
+	// re-authenticate to ensure we always have a fresh token
+	err := wait.PollUntilContextTimeout(ctx, 5*time.Second, 30*time.Second, true, func(ctx context.Context) (bool, error) {
+		err := kcClient.AuthenticatePassword(oidcClientId, "", "admin", "password")
+		return err == nil, err
+	})
+	require.NoError(t, err)
 
-	auth, err = tc.configClient.ConfigV1().Authentications().Get(ctx, "cluster", metav1.GetOptions{})
-	if err != nil {
-		return
-	}
+	group := names.SimpleNameGenerator.GenerateName("e2e-keycloak-group-")
+	err = kcClient.CreateGroup(group)
+	require.NoError(t, err)
 
-	origSpec := auth.Spec.DeepCopy()
-	auth.Spec.Type = configv1.AuthenticationTypeOIDC
-	auth.Spec.WebhookTokenAuthenticator = nil
-	auth.Spec.OIDCProviders = []configv1.OIDCProvider{
-		{
-			Name: idpName,
-			Issuer: configv1.TokenIssuer{
-				URL:       idpURL,
-				Audiences: []configv1.TokenAudience{oidcAudience},
-				CertificateAuthority: configv1.ConfigMapNameReference{
-					Name: oidcCABundleConfigMap,
-				},
-			},
-			ClaimMappings: configv1.TokenClaimMappings{
-				Groups: configv1.PrefixedClaimMapping{
-					TokenClaimMapping: configv1.TokenClaimMapping{
-						Claim: oidcGroupsClaim,
-					},
-				},
-				Username: configv1.UsernameClaimMapping{
-					TokenClaimMapping: configv1.TokenClaimMapping{
-						Claim: oidcUsernameClaim,
-					},
-				},
-			},
-			OIDCClients: []configv1.OIDCClientConfig{
-				{
-					ClientID:           "console",
-					ClientSecret:       configv1.SecretNameReference{Name: "console-secret"},
-					ComponentName:      "console",
-					ComponentNamespace: "openshift-console",
-				},
-			},
+	user := names.SimpleNameGenerator.GenerateName("e2e-keycloak-user-")
+	email := fmt.Sprintf("%s@test.dev", user)
+	password := "password"
+	firstName := "Homer"
+	lastName := "Simpson"
+	err = kcClient.CreateUser(
+		user,
+		email,
+		password,
+		[]string{group},
+		map[string]string{
+			"firstName": firstName,
+			"lastName":  lastName,
 		},
+	)
+	require.NoError(t, err)
+
+	httpClient := &http.Client{Transport: &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	}}
+
+	formData := url.Values{
+		"grant_type": []string{"password"},
+		"client_id":  []string{oidcClientId},
+		"scope":      []string{"openid"},
+		"username":   []string{user},
+		"password":   []string{password},
```

Comment:
Do we need to use an http client to post this form? Would it be reasonable to use a new keycloak client for the newly created user to get the token?